File: create_account/src/main/java/io/ebanking/application/account/AccountServer.java
package io.ebanking.application.account;

import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.stub.StreamObserver;
import java.io.IOException;
import java.util.concurrent.TimeUnit;

import org.mindrot.jbcrypt.BCrypt;
import java.sql.DriverManager;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.PreparedStatement;
import org.apache.commons.codec.digest.DigestUtils;

import javax.jmdns.JmDNS;
import javax.jmdns.ServiceEvent;
import javax.jmdns.ServiceListener;
import javax.jmdns.ServiceInfo;
import java.net.InetAddress;
import java.nio.file.Paths;

//https://github.com/grpc/grpc-java/blob/v1.31.1/examples/src/main/java/io/grpc/examples/helloworld/HelloWorldServer.java
//Service Discovery 
public class AccountServer {

  private static class AccountListener implements ServiceListener {
    @Override
    public void serviceAdded(ServiceEvent event) {
    }

    @Override
    public void serviceRemoved(ServiceEvent event) {
    }

    @Override
    public void serviceResolved(ServiceEvent event) {
    }
}

  private Server server;
  private void start() throws IOException {
      /* port used by server*/
    int port = 50051;
    server = ServerBuilder.forPort(port)
        .addService(new CreaterImpl())
        .build()
        .start();
    Runtime.getRuntime().addShutdownHook(new Thread() {
      @Override
      public void run() {
        
        System.err.println("Shuting Down Server");
        try {
          AccountServer.this.stop();
        } catch (InterruptedException e) {
          e.printStackTrace(System.err);
        }
        System.err.println("Shut Down Server");
      }
    });
  }

  private void stop() throws InterruptedException {
    if (server != null) {
      server.shutdown().awaitTermination(30, TimeUnit.SECONDS);
    }
  }
// Server Shutdown
  private void blockUntilShutdown() throws InterruptedException {
    if (server != null) {
      server.awaitTermination();
    }
  }
  /**
   
      Service Registration - Launch server
   */
  public static void main(String[] args) throws IOException, InterruptedException {
    final AccountServer server = new AccountServer();
    // Create JmDNS instance
    JmDNS jmdns = JmDNS.create(InetAddress.getLocalHost());

    // Service Registration -   Add a service listener
    ServiceInfo serviceInfo = ServiceInfo.create("_http._tcp.local.", "account", 50051, "path=/");
    jmdns.registerService(serviceInfo);
    jmdns.addServiceListener("_http._tcp.local.", new AccountListener());


    server.start();
    server.blockUntilShutdown();
  }

  static class CreaterImpl extends CreaterGrpc.CreaterImplBase {

    public boolean validString(String string) {
      for (int i = 0; i < string.length(); i++) {
        int c = (int) string.charAt(i);
        if(c == 34 || c == 10 || c == 8 || c == 39 || c == 13 || c == 9 || c == 92) {
          return false;
        }
      }
      return true;
    }

    @Override
    public void createAccount(AccountRequest req, StreamObserver<AccountReply> responseObserver) {
            
      Connection conn = null;
      String email = req.getEmail();
      String accountType = req.getAccount();
      String password = req.getPassword();


      if(email == null || email.equals("")) {
        AccountReply reply = AccountReply.newBuilder().setMessage("Please Specify an Email Address").setEmail(email).build();
        responseObserver.onNext(reply);
        responseObserver.onCompleted();
        return;
      }

      if(accountType == null || accountType.equals("")) {
        AccountReply reply = AccountReply.newBuilder().setMessage("Please Specify an Account Type").setAccountType(accountType).build();
        responseObserver.onNext(reply);
        responseObserver.onCompleted();
        return;
      }

      if(password == null || password.equals("")) {
        AccountReply reply = AccountReply.newBuilder().setMessage("Please Specify a Password").setPassword(password).build();
        responseObserver.onNext(reply);
        responseObserver.onCompleted();
        return;
      }

      email = email.toLowerCase();
      accountType = accountType.toLowerCase();
      password = DigestUtils.sha1Hex(password);
      // Needed for credentials
      String regex = "^[\\w!#$%&'*+/=?`{|}~^-]+(?:\\.[\\w!#$%&'*+/=?`{|}~^-]+)*@(?:[A-Z0-9-]+\\.)+[A-Z]{2,6}$".toLowerCase();
      if(!email.matches(regex) || !validString(email)) {
        AccountReply reply = AccountReply.newBuilder().setMessage("Please Specify a Valid Email Address").setEmail(email).build();
        responseObserver.onNext(reply);
        responseObserver.onCompleted();
        return;
      }

      if(!validString(accountType)) {
        AccountReply reply = AccountReply.newBuilder().setMessage("Please Specify a Valid Account Type").setAccountType(accountType).build();
        responseObserver.onNext(reply);
        responseObserver.onCompleted();
        return;
      }
      
      if(!validString(password)) {
        AccountReply reply = AccountReply.newBuilder().setMessage("Please Specify a Valid Password").setPassword(password).build();
        responseObserver.onNext(reply);
        responseObserver.onCompleted();
        return;
      }

        try {
            Class.forName("org.sqlite.JDBC");
            String url = "jdbc:sqlite:" + Paths.get("..").toAbsolutePath().normalize().toString() + "/ebanking.db";
            conn = DriverManager.getConnection(url);

            PreparedStatement query = conn.prepareStatement("SELECT accounts.accountType as accountType, users.user_id as user_id, users.password as password FROM users LEFT JOIN accounts ON users.user_id = accounts.user_id where email = ?");
            query.setString(1, email);
            ResultSet rs = query.executeQuery();
            boolean validAccount = true;
            int rowCount = 0;
            int user_id = 0;
            String databasePassword = "";
            if(rs != null) {
              while(rs.next()) {
                  
                  if(rs.getString("accountType") != null && rs.getString("accountType").equals(accountType)) {
                    validAccount = false;
                  }
                  user_id = rs.getInt("user_id");
                  databasePassword = rs.getString("password");
                  rowCount++;
              }
            }

            String responseMessage = "";

            if(rowCount <= 0) {

              query = conn.prepareStatement("INSERT INTO users (email, password) VALUES(?,?)");
              query.setString(1, email);
              query.setString(2, password);
              query.executeUpdate();

              query = conn.prepareStatement("SELECT users.user_id as user_id FROM users where email = ?");
              query.setString(1, email);
              rs = query.executeQuery();

              while(rs.next()) {
                  user_id = rs.getInt("user_id");
              }
              databasePassword = password;
              responseMessage = "A User with an Email Address of" + email + " has been created";
            }

            if (validAccount && user_id != 0) {

              if(!databasePassword.equals(password)) {
                AccountReply reply = AccountReply.newBuilder().setMessage("Invalid Credentials").setEmail(email).setAccountType(accountType).setPassword(password).build();
                responseObserver.onNext(reply);
                responseObserver.onCompleted();
                conn.close();
                return;
              }
              
              query = conn.prepareStatement("INSERT INTO accounts (accountType, user_id) VALUES(?,?)");
              query.setString(1, accountType);
              query.setInt(2, user_id);
              query.executeUpdate();
              if(!responseMessage.equals("")) {
                responseMessage = "A User with an Email Address of " + email + " with a " + accountType + " Account has been created";
              } else {
                responseMessage = "A " + accountType + " Account has been created for " + email;
              }
            }
            conn.close();

            if(!responseMessage.equals("")) {
              AccountReply reply = AccountReply.newBuilder().setMessage(responseMessage).setEmail(email).setAccountType(accountType).build();
              responseObserver.onNext(reply);
              responseObserver.onCompleted();
            } else {
              AccountReply reply = AccountReply.newBuilder().setMessage("A " + accountType + " Account already exists using the " + email + " Email Address").setEmail(email).setAccountType(accountType).build();
              responseObserver.onNext(reply);
              responseObserver.onCompleted();
            }

          } catch (Exception e) {
            e.printStackTrace();
            AccountReply reply = AccountReply.newBuilder().setMessage("An Error has Occured").setEmail(email).setAccountType(accountType).setPassword(password).build();
            responseObserver.onNext(reply);
            responseObserver.onCompleted();
        }
    }
  }
}

File: part-b/corba-redesign/account-service-redesign/account_server.py
import sys
import os
import sqlite3
import hashlib
import re

import CORBA, Bank, Bank__POA

class BankingServer_i(Bank__POA.BankingServer):
    def create_account(self, email, password, account):
        def valid_string(string):
            for i in range(len(string)):
                c = ord(string[i])
                if(c == 34 or c == 10 or c == 8 or c == 39 or c == 13 or c == 9 or c == 92):
                    return False;
            return True


        database_directory = os.path.join(os.path.dirname(os.path.realpath(__file__)), "../", "../", "ebanking.db")

        if(email == None or email == ""):
            return "Please specify an email address"

        if(account == None or account == ""):
            return "Please specify an account type"

        if(password == None or password == ""):
            return "Please specify a password"

        email = email.lower();
        account = account.lower();
        password = hashlib.sha1(password).hexdigest()

        regex = "^[\\w!#$%&'*+/=?`{|}~^-]+(?:\\.[\\w!#$%&'*+/=?`{|}~^-]+)*@(?:[A-Z0-9-]+\\.)+[A-Z]{2,6}$".lower()
        
        if((not re.match(regex, email)) or (not valid_string(email))):
            return "Please specify a valid email address"
      
        if(not valid_string(account)):
            return "Please specify a valid account type"
      
        if(not valid_string(password)):
            return "Please specify a valid password"

        try:

            conn = sqlite3.connect(database_directory)
            cursor = conn.cursor()
            sqlite_select_query = """SELECT accounts.accountType as accountType, users.user_id as user_id, users.password as password FROM users LEFT JOIN accounts ON users.user_id = accounts.user_id where email = "{email}";""".format(email = email, account = account)
            cursor.execute(sqlite_select_query)
            records = cursor.fetchall()

            row_count = 0
            user_id = 0
            database_password = None
            valid_account = True
            responseMessage = ""

            for i in records:
                database_account, user_id, database_password = i
                row_count += 1
                if database_account == account:
                  valid_account = False

            newly_created = False
            records = []
            if(row_count <= 0):
                cursor.execute(
                    """INSERT INTO users (email, password) VALUES("{email}","{password}");""".format(email = email, password = password)
                )
                conn.commit()
                newly_created = True
                cursor.execute("""SELECT users.user_id as user_id FROM users where email = "{email}";""".format(email = email))
                records = cursor.fetchall()
                if len(records) > 0:
                    user_id = records[0][0]
            
            account_created = False
            if valid_account and user_id != 0:
                if database_password != password and (not newly_created):
                    cursor.close()
                    return "Invalid credentials"
                cursor.execute(
                    """INSERT INTO accounts (accountType, user_id) VALUES("{account}","{user_id}");""".format(account = account, user_id = user_id)
                )
                conn.commit()
                account_created = True


            cursor.close()
            if newly_created and account_created:
                return "A user with an email address of " + str(email) + " with a " + str(account) + " account has been created"
            elif account_created:
                return "A " + str(account) + " account has been created for " + str(email)
            elif newly_created:
                return "A user with an email address of " + str(email) + " has been created"
            else:
                return "A " + str(account) + " account already exists using the " + str(email) + " email address"

        except Exception as e:
            print(e)
            return "An error occured"
    

orb = CORBA.ORB_init(sys.argv)
poa = orb.resolve_initial_references("RootPOA")

servant = BankingServer_i(  )
poa.activate_object(servant)

with open("ild_key.txt", "w") as f:
    f.write(orb.object_to_string(servant._this()))


poa._get_the_POAManager().activate(  )
orb.run(  )